Mathlib/CategoryTheory/Limits/IndYoneda.lean

```diff
@@ -0,0 +1,192 @@
+/-
+Copyright (c) 2024 Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christian Merten
+-/
+import Mathlib.CategoryTheory.Limits.Opposites
+import Mathlib.CategoryTheory.Limits.Preserves.Limits
+import Mathlib.CategoryTheory.Limits.Yoneda
+
+/-!
+
+# Ind- and pro- (co)yoneda lemmas
+
+We define limit versions of the yoneda and coyoneda lemmas.
+
+## Main results
+
+Notation: categories `C`, `I` and functors `D : Iᵒᵖ ⥤ C`, `F : C ⥤ Type`.
+
+- `colimitCoyonedaHomIsoLimit`: pro-coyoneda lemma: homorphisms from colimit of coyoneda of
+  diagram `D` to `F` is limit of `F` evaluated at `D`.
+- `colimitCoyonedaHomIsoLimit'`: a variant of `colimitCoyonedaHomIsoLimit` for a covariant
+  diagram.
+
+## TODO:
+
+- define the ind-yoneda versions (for contravariant `F`)
+
+-/
+
+universe u₁ u₂ v₁ v₂
+
+namespace CategoryTheory
+
+namespace Limits
+
+open Opposite
+
+variable {C : Type u₁} [Category.{u₂} C] {I : Type v₁} [Category.{v₂} I]
+
+section LimitOpColimit
+
+instance (F : I ⥤ C) [HasColimit F] : HasLimit F.op := hasLimit_op_of_hasColimit F
+
+/-- The limit of `F.op` is the opposite of `colimit F`. -/
+noncomputable def limitOpIsoOpColimit (F : I ⥤ C) [HasColimit F] :
+    limit F.op ≅ op <| colimit F :=
+  limit.isoLimitCone ⟨(colimit.cocone F).op, (colimit.isColimit F).op⟩
+
+@[simp]
+lemma limitOpIsoOpColimit_inv_comp_π (F : I ⥤ C) [HasColimit F] (i : I) :
+    (limitOpIsoOpColimit F).inv ≫ limit.π F.op ⟨i⟩ = (colimit.ι F i).op := by
+  simp [limitOpIsoOpColimit]
+
+instance (F : Iᵒᵖ ⥤ C) [HasColimit F] : HasLimit F.rightOp where
+  exists_limit := ⟨⟨coneRightOpOfCocone (colimit.cocone F),
+    isLimitConeRightOpOfCocone F (colimit.isColimit F)⟩⟩
+
+/-- `limitOpIsoOpColimit` for contravariant functor. -/
+noncomputable def limitRightOpIsoOpColimit (F : Iᵒᵖ ⥤ C) [HasColimit F] :
+    limit F.rightOp ≅ op <| colimit F :=
+  limit.isoLimitCone ⟨coneRightOpOfCocone (colimit.cocone F),
+    isLimitConeRightOpOfCocone F (colimit.isColimit F)⟩
+
+@[simp]
+lemma limitRightOpIsoOpColimit_inv_comp_π (F : Iᵒᵖ ⥤ C) [HasColimit F] (i : I) :
+    (limitRightOpIsoOpColimit F).inv ≫ limit.π F.rightOp i = (colimit.ι F ⟨i⟩).op := by
+  simp [limitRightOpIsoOpColimit]
+
+end LimitOpColimit
+
+section HomCocontinuousCovariant
+
+variable (F : I ⥤ C) [HasColimit F]
+
+/-- Hom is functorially cocontinuous: coyoneda of a colimit is the limit
+over coyoneda of the diagram. -/
+noncomputable def coyonedaOpColimitIsoLimitCoyoneda :
+    coyoneda.obj (op <| colimit F) ≅ limit (F.op ⋙ coyoneda) :=
+  coyoneda.mapIso (limitOpIsoOpColimit F).symm ≪≫ (preservesLimitIso coyoneda F.op)
+
+lemma coyonedaOpColimitIsoLimitCoyoneda_comp_π (i : I) :
```

Comment:
Because `by simp` did something (in implicit arguments) here, so I expected the simp NF linter to complain. I just tried to add it, let's see what the linter says.

Reply:
Linter is happy. Then my strategy of checking for simp NF with `by simp` is flawed.